python/eos/tasks.py

```diff
@@ -79,6 +81,97 @@ def task_wrapper(*args, **kwargs):
+    :type seed: int, optional
+    '''
+    if optimizations < 1:
+        raise ValueError('The number of optimizations should be larger than zero.')
+    if seed and not start_point is None:
+        raise ValueError('The arguments seed and start_point are mutually exclusive')
+    if seed and not from_mcmc is None:
+        raise ValueError('The arguments seed and fron_mcmc are mutually exclusive')
+    if not from_mcmc is None and not start_point is None:
+        raise ValueError('The arguments from_mcmc and start_point are mutually exclusive')
+
+    if type(analysis_file) is not eos.AnalysisFile:
+        _analysis_file = eos.AnalysisFile(analysis_file)
+    else:
+        _analysis_file = analysis_file
+    analysis = _analysis_file.analysis(posterior)
+    min_chi2 = sys.float_info.max
+    gof = None
+    bfp = None
+
+    eos.info(f'Starting minimization in {optimizations} points')
+
+    if not start_point is None:
+        starting_point = _np.array(start_point)
+        eos.info('Starting optimization from user-provided point')
+
+        _bfp = analysis.optimize(start_point = starting_point)
+        _gof = eos.GoodnessOfFit(analysis._log_posterior)
+        _chi2 = _gof.total_chi_square()
+    elif from_mcmc >= 0:
+        eos.info('Initializing starting point from MCMC data file')
+        chain = eos.data.MarkovChain(os.path.join(base_directory, posterior, f'mcmc-{from_mcmc:04}'))
+        idx_mode = _np.argmax(chain.weights)
+        for p, v in zip(analysis.varied_parameters, chain.samples[idx_mode]):
+            p.set(v)
+
+        _bfp = analysis.optimize()
+        _gof = eos.GoodnessOfFit(analysis._log_posterior)
+        _chi2 = _gof.total_chi_square()
+    else:
+        eos.info('Starting optimization from a random point')
+        if seed is None:
+            seed = 17
+        _bfp = analysis.optimize(start_point='random', rng=_np.random.mtrand.RandomState(seed))
+        _gof = eos.GoodnessOfFit(analysis._log_posterior)
+        _chi2 = _gof.total_chi_square()
+
+    for i in range(optimizations - 1):
+        starting_point = [float(p) for p in analysis.varied_parameters]
+        _bfp = analysis.optimize(start_point = starting_point)
+        _gof = eos.GoodnessOfFit(analysis._log_posterior)
+        _chi2 = _gof.total_chi_square()
+        if _chi2 < min_chi2:
+            gof = _gof
+            bfp = _bfp
+            min_chi2 = _chi2
+
+    print('best-fit point:')
+    for p, v in zip(analysis.varied_parameters, _bfp.point):
+        print('  - {} -> {}'.format(p.name(), v))
+    print('total chi^2 = {}'.format(min_chi2))
+    print('p value = {:.1f}%'.format(100 * (1.0 - scipy.stats.chi2(gof.total_degrees_of_freedom()).cdf(gof.total_chi_square()))))
+    print('individual test statistics:')
+    for n, e in gof:
+        print('  - {}: chi^2 / dof = {:f} / {}'.format(n, e.chi2, e.dof))
```

Comment:
This should now be written into a YAML file in ``{posterior}/modes``

Reply:
Also, should this be **appended** to that file? ``modes`` implies more than one mode. If yes, how do we avoid stale modes lingering past their lifetime?

Reply:
You can indeed have several modes depending on your starting_point. Doing this exercise correctly would require keeping all the putative bfp, comparing them to see if they really are different modes etc... Is that what we want?

Reply:
For this task, I would equate BFP with mode. Distinguishing would not be needed here. I'm just concerned that several calls to ``find-mode`` with changes to the analysis inbetween could make a mess...

Reply:
If the starting_point is fixed, it shouldn't, right?

Reply:
I am concerned about incompatible analyses. E.g., changing the likelihood, or changing the dimensionality of the parameter space.

Reply:
@mreboud and I discussed offline and agreed to label the outputs. Modes will be written into ``POSTERIOR/mode-LABEL``, where the label is user configurable and defaults to ``default``.